### PR TITLE
Map Generator

### PR DESCRIPTION
**Pull request by TrueBabyChaise**
Basic map and ColliderComponent, system hasn't be done at the moment

### CODE REVIEW

File: sources/Map/Map.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	{
+		std::map<BlockType, std::function<void (Vector3f coords, Vector3f size, WAL::Scene &)>> elements = {
+			{BREAKABLE, &createBreakable},
+			{UNBREAKABLE, &createUnbreakable},
+			{HOLE, createHole},
+			{BUMPER, createBumper},
+			{STAIRS, createStairs}
+		};
+		
+		for (auto element : elements)
```
Comment (AnonymusRaccoon) [inline sources/Map/Map.cpp:30]: You should use std::find or std::find_if

File: sources/Map/Map.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	static void Map::generateMap(int width, int height, int seed, WAL::Scene &scene)
+	{
+		std::map<std::tuple<int, int>, char> map;
+		
+		for (int i = 0; i < width; i++)
+			for (int j = 0; j < height; j++)
+				map[std::make_tuple(i, j)] = ' '
+		for (int i = 0; i < 4; i++)
+			map[std::make_tuple((int)(std::pow(seed, i) % (width - 1) + 1),(int)(std::pow(seed * 0.7, i) % (height - 1) + 1)] = '*';
+		for (int i = 1; i < width - 1; i++)
```
Comment (AnonymusRaccoon) [inline sources/Map/Map.cpp:131]: Put brackets arrround nested for/ifs

File: sources/Map/Map.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+			case STAIRS:
+				return 'S';
+			default:
+				return ' ';
+		}
+	}
+
+	static void Map::generateMap(int width, int height, int seed, WAL::Scene &scene)
+	{
+		std::map<std::tuple<int, int>, char> map;
```
Comment (AnonymusRaccoon) [inline sources/Map/Map.cpp:124]: This is never queried.

File: sources/Map/Map.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+			for (int j = 0; j < height; j++)
+				map[std::make_tuple(i, j)] = ' '
+		for (int i = 0; i < 4; i++)
+			map[std::make_tuple((int)(std::pow(seed, i) % (width - 1) + 1),(int)(std::pow(seed * 0.7, i) % (height - 1) + 1)] = '*';
+		for (int i = 1; i < width - 1; i++)
+			for (int j = 1; j < height - 1; j++)
+				if (isBlockCloseToPlayer(map[std::make_tuple(i, j)], i , j))
+					map[std::make_tuple(i, j)] = ' ';
+				else
+					map[std::make_tuple(i, j)] = enumToChar(getRandomBlockType(seed, i * width + j)
```
Comment (AnonymusRaccoon) [inline sources/Map/Map.cpp:136]: Why are you using chars instead of the enum type directly?

File: sources/Map/Map.hpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#include <tuple>
+
+namespace BBM
+{
+
+	class Map
+	{
+		private:
+			//! @brief Enum of the block available.
+			static enum BlockType {
```
Comment (AnonymusRaccoon) [inline sources/Map/Map.hpp:21]: Why are you making a static enum? An enum is enough

**Review by AnonymusRaccoon — COMMENTED** [review]
There is already a draft branch with a ColliderComponent, please don't duplicate work

File: sources/Map/Map.cpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+	static void Map::generateMap(int width, int height, int seed, WAL::Scene &scene)
+	{
+		std::map<std::tuple<int, int>, char> map;
+		
+		for (int i = 0; i < width; i++)
+			for (int j = 0; j < height; j++)
+				map[std::make_tuple(i, j)] = ' '
+		for (int i = 0; i < 4; i++)
+			map[std::make_tuple((int)(std::pow(seed, i) % (width - 1) + 1),(int)(std::pow(seed * 0.7, i) % (height - 1) + 1)] = '*';
```
Comment (Octopus773) [inline sources/Map/Map.cpp:130]: C cast in C++ are a big no no
use static_cast\<Type\>(value)
here:
```suggestion
			map[std::make_tuple(static_cast<int>(std::pow(seed, i) % (width - 1) + 1),  static_cast<int>(std::pow(seed * 0.7, i) % (height - 1) + 1)] = '*';
```

Comment (TrueBabyChaise) [reply sources/Map/Map.cpp]: it's mistake, we'll change right now.

File: sources/Map/Map.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		scene->addEntity("Hole Block")
+			.addComponent<PositionComponent>(coords)
+			.addComponent<Drawable3DComponent<RAY3D::Model>>("assets/wall/hole_block.obj", std::make_pair(MAP_DIFFUSE, "assets/wall/hole_block.png"));
+			/* .addComponent<ColliderComponent>([](const WAL::Entity &entity, WAL::Entity &other) {
+				if (other.hasComponent<HealthComponent>()) {
+					auto &health = other.getComponent<HealthComponent>();
+					health.takeDmg(health.getHealthPoint());
+				}
+			}); */
+		WAL::Entity entity("");
```
Comment (Octopus773) [inline sources/Map/Map.cpp:100]: please suppr this var

File: sources/Map/Map.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+			map[std::make_tuple(x + 1, y)] == blockType ||
+			map[std::make_tuple(x, y + 1)] == blockType ||
+			map[std::make_tuple(x, y - 1)] == blockType)
+			return (true);
+		return (false);
+	}
+	
+	Map::BlockType Map::getRandomBlockType(int seed, int blockCreated)
+	{
+		return static_cast<BlockType>((seed * blockCreated * rand()) % (HOLE));
```
Comment (Octopus773) [inline sources/Map/Map.cpp:133]: this function should not take any args

File: sources/Map/Map.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	{
+		std::map<std::tuple<int, int>, BlockType> map;
+		
+		for (int i = 0; i < width; i++)
+			for (int j = 0; j < height; j++)
+				map[std::make_tuple(i, j)] = NOTHING;
+		map[std::make_tuple(1, 1)] = SPAWNER;
+		map[std::make_tuple(width - 1, 1)] = SPAWNER;
+		map[std::make_tuple(1, height - 1)] = SPAWNER;
+		map[std::make_tuple(width - 1, height - 2)] = SPAWNER;
```
Comment (GitBluub) [inline sources/Map/Map.cpp:146]: Height - 1 or Height - 2 for the bottom spawners ?

File: sources/Map/Map.cpp [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+namespace RAY3D = RAY::Drawables::Drawables3D;
+
+namespace BBM
+{
+	void Map::generateWall(int width, int height, std::shared_ptr<WAL::Scene> scene)
+	{
+		for (int i = 0; i < width; i++) {
+			scene->addEntity("Width Wall")
+				.addComponent<PositionComponent>(Vector3f(i,0,height))
+				.addComponent<Drawable3DComponent<RAY3D::Model>>("assets/wall/unbreakable_wall.obj", std::make_pair(MAP_DIFFUSE, "assets/wall/unbreakable_wall.png"));
```
Comment (Octopus773) [inline sources/Map/Map.cpp:19]: you should use var for often used paths

**Comment by Octopus773** [comment]
player spawn isn't set correctly

**Comment by Octopus773** [comment]
you should call std::srand

**Pull request closed without merging**